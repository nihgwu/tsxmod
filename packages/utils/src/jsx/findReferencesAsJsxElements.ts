import {
  Identifier,
  JsxOpeningElement,
  JsxSelfClosingElement,
  Node,
} from 'ts-morph'

/**
 * Traces component references.
 *
 * This is similar to `findReferencesAsNodes` but returns JsxSelfClosingElement and JsxElement nodes.
 * Note, this currently does not account for cases where the component is used as a prop or is renamed.
 */
export function findReferencesAsJsxElements(
  identifer: Identifier
): (JsxOpeningElement | JsxSelfClosingElement)[] {
  const jsxElements: Node[] = []

  for (const reference of identifer.findReferencesAsNodes()) {
    const node = reference.getFirstAncestor((node) => {
      return (
        Node.isJsxOpeningElement(node) || Node.isJsxSelfClosingElement(node)
      )
    })

    if (node) {
      jsxElements.push(node)
    }
  }

  return jsxElements as (JsxOpeningElement | JsxSelfClosingElement)[]
}
