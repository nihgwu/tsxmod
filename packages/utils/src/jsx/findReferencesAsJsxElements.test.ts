import { Project, SyntaxKind } from 'ts-morph'
import { findReferencesAsJsxElements } from './findReferencesAsJsxElements'

const buttonSource = `
import React from 'react'

export const Button = (props: {
  onClick: () => void
  children: React.ReactNode
}) => {
  return <button onClick={props.onClick}>{props.children}</button>
}
`

const dialogSource = `
import React from 'react'
import { Button } from './Button'

export function Dialog(props: {
  open: boolean
  onClose: () => void
  children: React.ReactNode
}) {
  if (!props.open) return null

  return (
    <dialog>
      {props.children}
      <Button onClick={props.onClose}>Close</Button>
    </dialog>
  )
}
`

const appSource = `
import React from 'react'
import { Button } from './Button'
import { Dialog } from './Dialog'

export default function App() {
  const [open, setOpen] = React.useState(false)

  return (
    <>
      <Button onClick={() => setOpen(true)} children="Open" />
      <Dialog onClose={() => setOpen(false)} open={open} />
    </>
  )
}
`

test('finds all element references to a component', () => {
  const project = new Project({ useInMemoryFileSystem: true })

  const buttonSourceFile = project.createSourceFile('Button.tsx', buttonSource)

  project.createSourceFile('Dialog.tsx', dialogSource)
  project.createSourceFile('App.tsx', appSource)

  const componentIdentifier = buttonSourceFile
    .getVariableDeclarationOrThrow('Button')
    .getFirstDescendantByKindOrThrow(SyntaxKind.Identifier)
  const references = findReferencesAsJsxElements(componentIdentifier)

  expect(references.map((reference) => reference.getText())).toMatchSnapshot()
})
