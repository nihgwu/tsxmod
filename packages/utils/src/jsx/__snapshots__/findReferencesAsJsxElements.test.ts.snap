// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds all element references to a component 1`] = `
[
  "<Button onClick={props.onClose}>",
  "<Button onClick={() => setOpen(true)} children="Open" />",
]
`;
